### PR TITLE
Add an option to filter reads on the header filter flag

### PR DESCRIPTION
**Pull request by kstromhaug**
Added a --keep-only-passing flag to keep only reads annotated with Y in the header comment.

### CODE REVIEW

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:151]: bingo!
```suggestion
```

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:298]: needs more description of what a "passing filter flag" is and where it comes from (i.e. the `<is filtered>` field in the FASTQ comment above).

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:349]:
```suggestion
 @arg(doc="Keep only passing filter reads if true, otherwise keep all reads. Passing filter reads are determined from the comment in the FASTQ header.")
```

**Review by nh13 — CHANGES_REQUESTED** [review]
Can you add a bit more in the usage about the new option?

Once the #711 PR is merged, we'll rebase and retarget this PR to the `master` branch before merging.  Approvals needed of course.

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:776]: @kstromhaug @tfenne part of me wonders if we should check `keepOnlyPassing` here when setting `passQc` on the `DemuxResult`, but instead check it when we filter the `resultIterator`.

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:775]: Then update `resultsIterator.filter`
```suggestion
    val passQc = demuxRecords.forall(d => d.readInfo.forall(_.passQc))
```

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:168]:
```suggestion
    resultIterator.filter(r => r.keep(keepOnlyPassing))
```

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:627]: Add a doc-string :)
```suggestion
    /** Returns true if this [[DemuxResult]] should be kept for output, false otherwise.
     *
     * Returns true if `keepOnlyPassing` is false or if `passQc` is true
     *
     * @param keepOnlyPassing true to keep only passing reads, false to keep all reads
    */
    def keep(keepOnlyPassing: Boolean = false): Boolean = {
```

**Review by nh13 — APPROVED** [review]
One minor docstring that I'd like you to add, but LGTM.  Please wait for @tfenne's review, and #711 to merge into main before proceeding.

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:350]: @kstromhaug @tfenne  do you think it's clearer to have this option named `--omit-failing-reads` with default `false` (outputs all reads), and when `true` only emits passing reads determined from the FASTQ header/comment?

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:298]: This sounds to me like it's mixing two things - the `passing filter` flag and the `is control` flag/code.  How about

```suggestion
      |By default all input reads are output.  If your input FASTQs contain reads that do not pass filter (as defined by the Y/N filter flag in the FASTQ comment) these can be filtered out during demultiplexing using the `--skip-failing-reads` option.
```

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala
Comment (tfenne) [reply src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:350]: I would definitely rename this to be something like `--skip-failing-reads` or `--omit-failing-reads` as I think that is clearer.

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:791]: Wouldn't it be a lot more efficient to do this up front?  What's the point of demuxing reads that we're just going to throw on the floor?  I would suggest doing this wherever `demultiplex()` is called from and then just skipping over the input fastq records.

**Review by tfenne — CHANGES_REQUESTED** [review]
A couple of minor things, but one more major comment/suggestion about moving the logic to skip failing reads up out of `demultiplex()` so that we don't spend a lot of effort demuxing reads we don't output.

I guess the other question is if you say to only do passing reads, do you want stats for non-passing reads?  I would think not?

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala
Comment (kstromhaug) [reply src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:789]: Leaving as is for now

**Review by tfenne — APPROVED** [review]

**Pull request merged**
